insights/tests/test_lazylogfileoutput.py

```diff
@@ -0,0 +1,362 @@
+     profile = manager.get_profile(consumer_id, content_type)
+   File "/usr/lib/python2.6/site-packages/pulp/server/managers/consumer/profile.py", line 120, in get_profile
+     raise MissingResource(profile_id=profile_id)
+MissingResource: Missing resource(s): profile_id={'content_type': u'rpm', 'consumer_id': u'1786cd7f-2ab2-4212-9798-c0a454e97900'}
+Mar 27 03:39:43 system rsyslogd-2177: imuxsock begins to drop messages from pid 55082 due to rate-limiting
+Mar 27 03:39:46 system rsyslogd-2177: imuxsock lost 165 messages from pid 55082 due to rate-limiting
+Mar 27 03:39:49 system rsyslogd-2177: imuxsock begins to drop messages from pid 55082 due to rate-limiting
+Mar 27 03:49:10 system pulp: pulp.server.webservices.middleware.exception:ERROR: Missing resource(s): profile_id={'content_type': u'rpm', 'consumer_id': u'79d5aed1-5631-4f40-b970-585ee974eb87'}
+"""
+
+
+def test_lazylogfileoutput_scanners():
+    # Messages that are present can be kept
+    FakeMessagesClass.keep_scan('puppet_master_logs', ' puppet-master')
+    # Messages that are absent still turn up as an empty list
+    FakeMessagesClass.keep_scan('kernel_logs', ' kernel')
+    # Token scan of something that's present
+    FakeMessagesClass.token_scan(
+        'middleware_exception_present', 'pulp.server.webservices.middleware.exception'
+    )
+    # Token scan of something that's absent
+    FakeMessagesClass.token_scan('cron_present', 'CRONTAB')
+
+    # Check that duplicate scanners raise a value exception
+    with pytest.raises(ValueError) as exc:
+        FakeMessagesClass.keep_scan('kernel_logs', ' kernel')
+    assert "'kernel_logs' is already a registered scanner key" in str(exc)
+
+    def count_lost_messages(self):
+        imuxsock_lines = 0
+        lost_messages = 0
+        for line in self.lines:
+            if 'imuxsock lost' in line:
+                parts = line.split(None)
+                if parts[7].isdigit():
+                    imuxsock_lines += 1
+                    lost_messages += int(parts[7])
+        return "lost {msgs} messages in {lines} lines".format(
+            msgs=lost_messages, lines=imuxsock_lines
+        )
+
+    FakeMessagesClass.scan('lost_messages', count_lost_messages)
+
+    ctx = context_wrap(MESSAGES, path='/var/log/messages')
+    log = FakeMessagesClass(ctx)
+    assert not hasattr(log, 'puppet_master_logs')
+    assert not hasattr(log, 'kernel_logs')
+    assert not hasattr(log, 'middleware_exception_present')
+    assert not hasattr(log, 'cron_present')
+    assert not hasattr(log, 'lost_messages')
+    log.do_scan('puppet_master_logs')
+    assert hasattr(log, 'puppet_master_logs')
+    log.do_scan('kernel_logs')
+    assert hasattr(log, 'kernel_logs')
+    log.do_scan('middleware_exception_present')
+    assert hasattr(log, 'middleware_exception_present')
+    log.do_scan('cron_present')
+    assert hasattr(log, 'cron_present')
+    log.do_scan('lost_messages')
+    assert hasattr(log, 'lost_messages')
```

Comment:
[nitpick] 
The codecov reports for [partial coverage lines](https://app.codecov.io/gh/RedHatInsights/insights-core/pull/4309?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=RedHatInsights#e3ed9166e35444a62975de12bf0d55b7-R1429) on:
```
+     def do_scan(self, result_key=None):
+         if result_key:
+             if result_key not in self._scanned and result_key in self.scanners:     <-- this line
+                 self.scanners[result_key](self)
+                 self._scanned.add(result_key)
+         else:
+             for key, scanner in self.scanners.items():
+                 if key not in self._scanned:                                        <-- and this line
```

@xiangce  Would it be possible to add a few more test lines to trigger this `key in self._scanned`?  And `result_key not in self.scanners` if needed?

Reply:
Sure things, will add them

Reply:
Done, but the Codecov cannot sync the result back to the PR so quickly. It's covered, see https://app.codecov.io/gh/RedHatInsights/insights-core/pull/4309?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=RedHatInsights#e3ed9166e35444a62975de12bf0d55b7-R1429